pkg/config/cfgvars.go

```diff
@@ -0,0 +1,303 @@
+		BinDir:                     filepath.Join(dataDir, "bin"),
+		OCIBundleDir:               filepath.Join(dataDir, "images"),
+		CertRootDir:                certDir,
+		DataDir:                    dataDir,
+		EtcdCertDir:                filepath.Join(certDir, "etcd"),
+		EtcdDataDir:                filepath.Join(dataDir, "etcd"),
+		KineSocketPath:             filepath.Join(runDir, constant.KineSocket),
+		KonnectivitySocketDir:      filepath.Join(runDir, "konnectivity-server"),
+		KubeletAuthConfigPath:      filepath.Join(dataDir, "kubelet.conf"),
+		KubeletVolumePluginDir:     constant.KubeletVolumePluginDir,
+		ManifestsDir:               filepath.Join(dataDir, "manifests"),
+		RunDir:                     runDir,
+		KonnectivityKubeConfigPath: filepath.Join(certDir, "konnectivity.conf"),
+		RuntimeConfigPath:          filepath.Join(runDir, "k0s.yaml"),
+		StatusSocketPath:           filepath.Join(runDir, "status.sock"),
+		StartupConfigPath:          constant.K0sConfigPathDefault,
+
+		// Helm Config
+		HelmHome:             helmHome,
+		HelmRepositoryCache:  filepath.Join(helmHome, "cache"),
+		HelmRepositoryConfig: filepath.Join(helmHome, "repositories.yaml"),
+
+		origin: CfgVarsOriginDefault,
+	}
+
+	if cmd != nil {
+		WithCommand(cmd)(vars)
+	}
+
+	return vars, nil
+}
+
+func (c *CfgVars) Cleanup() error {
+	if c.origin == CfgVarsOriginDefault && file.Exists(c.RuntimeConfigPath) {
+		return os.Remove(c.RuntimeConfigPath)
+	}
+	return nil
+}
+
+func (c *CfgVars) defaultStorageSpec() *v1beta1.StorageSpec {
+	if c.DefaultStorageType == v1beta1.KineStorageType {
+		return &v1beta1.StorageSpec{
+			Type: v1beta1.KineStorageType,
+			Kine: v1beta1.DefaultKineConfig(c.DataDir),
+		}
+	}
+
+	return v1beta1.DefaultStorageSpec()
+}
+
+func (c *CfgVars) NodeConfig() (*v1beta1.ClusterConfig, error) {
+	if c.nodeConfig != nil {
+		return c.nodeConfig, nil
+	}
+
+	if c.origin == CfgVarsOriginRuntime {
+		return nil, fmt.Errorf("runtime config is not available")
+	}
+
+	if c.StartupConfigPath == "" {
+		return nil, fmt.Errorf("config path is not set")
+	}
+
+	var nodeConfig *v1beta1.ClusterConfig
+
+	cfgContent, err := os.ReadFile(c.StartupConfigPath)
+	if errors.Is(err, os.ErrNotExist) && c.StartupConfigPath == constant.K0sConfigPathDefault {
+		nodeConfig = v1beta1.DefaultClusterConfig(c.defaultStorageSpec())
+	} else if err == nil {
+		cfg, err := v1beta1.ConfigFromString(string(cfgContent), c.defaultStorageSpec())
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig = cfg
+	} else {
+		return nil, err
+	}
+
+	if nodeConfig.Spec.Storage.Type == v1beta1.KineStorageType && nodeConfig.Spec.Storage.Kine == nil {
+		nodeConfig.Spec.Storage.Kine = v1beta1.DefaultKineConfig(c.DataDir)
+	}
+
+	c.nodeConfig = nodeConfig
+
+	return nodeConfig, nil
+}
+
+func (c *CfgVars) ClusterConfig(ctx context.Context, kubeClientFactory kubeutil.ClientFactoryInterface) (*v1beta1.ClusterConfig, error) {
```

Comment:
Since we're on it right now: What about renaming this function to `LoadClusterConfig`?

Reply:
I'd call it `FetchClusterConfig` or similar in that case as I'd like it to be contextually as far as possible from `loading` which is usually done to files.

Reply:
Renamed to `FetchDynamicConfig`